File: 430. Flatten a Multilevel Doubly Linked List.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, prev, next, child):
        self.val = val
        self.prev = prev
        self.next = next
        self.child = child
"""
class Solution:
    def flatten(self, head: 'Node') -> 'Node':
        if not head or (not head.next and not head.child):
            return head
        head, _ = self.helper(head)
        return head
            
            
    def helper(self, node):
        head = node
        while node:
            if not node.child:
                if not node.next:
                    break
                else:
                    node = node.next
            else:
                if node.next:
                    saved = node.next
                    chead, ctail = self.helper(node.child)
                    node.next = chead
                    chead.prev = node
                    saved.prev = ctail
                    ctail.next = saved
                    node.child = None
                    node = saved
                else:
                    chead, ctail = self.helper(node.child)
                    node.next = chead
                    chead.prev = node
                    node.child = None
                    node = ctail
                    break
            
        return head, node

File: 430. Flatten a Multilevel Doubly Linked List.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node prev;
    public Node next;
    public Node child;

    public Node() {}

    public Node(int _val,Node _prev,Node _next,Node _child) {
        val = _val;
        prev = _prev;
        next = _next;
        child = _child;
    }
};
*/
class Solution {
    public Node flatten(Node head) {
        if (head == null || (head.next == null && head.child == null)) {
            return head;
        }
        Node next = null, node = head, temp=null;
        while (node != null) {
            next = node.next;
            if (node.child != null) {
                temp = flatten(node.child);
                node.child = null;
                temp.prev = node;
                node.next = temp;
                while (temp.next != null) {
                    temp = temp.next;
                }
                if (next != null) {
                    temp.next = next;
                    next.prev = temp;
                }
            }
            node = next;
        }
        return head;
    }
}

class Solution {
    public Node flatten(Node head) {
        if (head == null || (head.next == null && head.child == null)) {
            return head;
        }
        Node next = null, node = head, temp=null;
        while (node != null) {
            next = node.next;
            if (node.child != null) {
                temp = node.child;
                node.next = temp;
                temp.prev = node;
                while (temp.next != null) {
                    temp = temp.next;
                }
                temp.next = next;
                if (next != null) {
                    next.prev = temp;
                }
                node.child = null;
            } else {
                node = node.next;
            }
        }
        return head;
    }
}